pytorch_lightning/core/lightning.py

```diff
@@ -263,9 +263,13 @@ def _apply_batch_transfer_handler(
         self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0
     ) -> Any:
         device = device or self.device
-        batch = self.on_before_batch_transfer(batch, dataloader_idx)
-        batch = self.transfer_batch_to_device(batch, device, dataloader_idx)
-        batch = self.on_after_batch_transfer(batch, dataloader_idx)
+        batch = self.trainer._data_connector._datahook_source.get_hook("on_before_batch_transfer")(
```

Comment:
Does it mean only the objects providing the data loaders can provide the implementation for those hooks?
I am fine with this change, but this would be behavior BC.

Reply:
I would say yes.. but we can make adjustments by checking where these hooks are overridden and using that one. In case the same hook is overridden in both Module and DataModule, then we can either raise an error or prioritize datamodule with a warning. wdyt?

Reply:
I am fine with this change, just wanted to be sure.